### ISSUE TITLE
adhere to the widespread comment directive format

### ISSUE THREAD

**Issue by mvdan** [golangci/golangci-lint#1658]
In Go, `//go:*` is reserved for the Go toolchain. See https://golang.org/cmd/compile/.

It is encouraged that other Go tools use a similar "namespace" prefix for their comment directives. For example, `//gccgo:*`, `//lint:*` for staticcheck, `//go-sumtype:*`, and so on. See https://groups.google.com/g/golang-dev/c/r4rdPdsH1Fg/m/tNZazPenX5cJ.

golangci-lint kind of does this, but not really. https://golangci-lint.run/usage/false-positives/#nolint gives this example:

> var bad_name int //nolint

That's no good, because it clearly does not adhere to the widespread format. It is not formally defined, but in practice it follows something like `^[a-z]+:[a-z]+`.

A possible fix would be `//nolint:all`, which is also clearer. Another option would be to follow staticcheck's https://staticcheck.io/docs/#line-based-linter-directives, which already follows the format with `//lint:ignore ...`.

**Comment by mvdan** [comment]
For the sake of documenting this better upstream, I've raised https://github.com/golang/go/issues/43776.

**Comment by kortschak** [comment]
Moving to the approach used by staticcheck would also have the advantage that file-wide directives will work properly; currently the way to do a file-wide directed is to attach the comment to the package decl, which in turn results in a line in the godoc.

See https://staticcheck.io/docs/#file-based-linter-directives for how staticcheck does it.

**Comment by kortschak** [comment]
Also, @ldez given that this is a deviation from the Go standard, I think this *is* a bug and not an enhancement.

**Comment by ldez** [comment]
Today, `nolint` can have to syntax
- just a simple comment: `// nolint`
- a directive syntax `//nolint`

In the code base of golangci-lint, it's used a simple comment syntax. So from my point of view, there is no deviation.

> currently the way to do a file-wide directed is to attach the comment to the package decl, which in turn results in a line in the godoc.

currently the `nolint` directives are ignored in the godoc https://github.com/golangci/golangci-lint/issues/1566#issuecomment-749094833

**Comment by mvdan** [comment]
> just a simple comment: `// nolint`

To be honest, I don't think this is a good solution. All golangci-lint directives should clearly look like directives, not plain English comments.

> there is no deviation.

I don't follow - the docs clearly point towards `//nolint`, which does not use a space and does not adhere to the format.

> the resulting breaking change

The fix could be rolled out over time, or come with an automated way to fix up older comments like https://github.com/ipld/specs/pull/353/files#diff-ed097ee1bca871446cf885a8fabf9722c54921cf7bae59f780cd14a6f305c3d5.

**Comment by AlekSi** [comment]
I would propose for golangci-lint to complain about all forms with spaces: `// nolint`, `//nolint: name`, etc.

**Comment by ldez** [comment]
@AlekSi you can already to that:
```yml
linters-settings:
  nolintlint:
    # Disable to ensure that nolint directives don't have a leading space. Default is true.
    allow-leading-space: true
```

**Comment by AlekSi** [comment]
And I think the default should be changed to `false` in some future version with some warnings in between.

**Comment by kortschak** [comment]
It would be very nice if golangci-lint respected the linter directives of supported linters.

```
$ staticcheck .                                                          
$ golangci-lint run    
main.go:7:6: `foo` is unused (deadcode)
func foo() {}
     ^
$ cat main.go   
package main

func main() {
}

//lint:ignore U1000 Included for demonstration purposes.
func foo() {}
```

**Comment by mvdan** [comment]
See the upstream bug report above; users are starting to run into this problem now that gofmt reformats godoc comments. I would say that fixing this bug before 1.19 is released in three weeks is very important - otherwise the number of incoming bug reports will multiply.

**Comment by ldez** [comment]
I will create a PR to drop `allow-leading-space` option and add the support of `nolint:all`.

It will not be the real fix for this issue but a fix for go1.19 gofmt.

In parallel, I started to work on a real fix for this issue with a new syntax.
We cannot use exactly the same option as staticcheck (`//lint:ignore`) because it will conflict, so I think we will use `//lint:skip linter1[,linter2,...,linterN] reason`.
This will be introduced in a v2 of golangci-lint.

**Comment by dominikh** [comment]
> We cannot use exactly the same option as staticcheck (`//lint:ignore`) because it will conflict

I explicitly designed the semantics of that directive to be usable by more tools than just Staticcheck (which is why the prefix is `lint:`). In particular, Staticcheck complains about unused ignore directives, but only if it actually has those checks enabled. That means other linters are free to use the same directive with their own check names.

**Comment by ldez** [comment]
I'm talking about the "option" part of the directive: `ignore`.

**Comment by dominikh** [comment]
I don't understand what the problem is? If you saw `//lint:ignore SA4000` you would ignore it, the same way that Staticcheck would ignore `//lint:ignore somelinter`. The two uses of `//lint:ignore` can coexist.

**Comment by mvdan** [comment]
I think Dominik means that golangci-lint should ignore directives that it doesn't understand, like `//lint:ignore SA4000`, because `SA4000` is not a linter name that it knows. This is how many directives in Go work: for example, the Go compiler obeys `go:noinline`, but will ignore `go:foobar`. And `encoding/json` will obey ``Field string `json:",omitempty"``, but will ignore ``Field string `json:",foobar"``. Not just for better interoperability with other implementations, but also for fowards compatibility with newer versions adding new directives.

**Comment by ldez** [comment]
For now, no strong decision, I'm just thinking about it.

I have to see what really needs to be kept from the past, and how the behavior can be extended to manage the specific constraints of golangci-lint.

The current implementation of `nolint` is too complicated due to the previous lax syntax, so I don't want to reproduce the same errors.

So I will not take a decision quickly, and currently, it's not my top priority.
I'm working a bit on several topics to be able to have the time to think about them in the background.

**Comment by mvrahden** [comment]
My linters are colliding already (see: https://github.com/mvdan/gofumpt/issues/241) and I do not believe that waiting for a v2 can actually be the go-to option here, except the v2 release is right around the corner.

Why not use the auto-fix option that was proposed previously in this thread?  
I think `lint:skip` sounds like a good plan to go forward with, there's no need to serve both linters staticcheck and golangci-lint under one directive. As people will either use golangci-lint linter (with staticcheck,unused,stylecheck... as a plugin) or staticcheck as standalone solution.

The fact that golangci-lint reports non-existing linters is actually a good thing, because you'll get feedback about the tool setup, otherwise you end up having nolint directives which will only perform half of what they are configured to be doing.

Anyway, golangci-lint is not the same use case as staticcheck here, because staticcheck maintains a catalogue of checks, which is less volatile than the linters being added and deprecated within this project. So it actually golangci-lint deserves this sanitization feature and therefore its configuration should live under its own directive domain.

```sh
PS: A workaround to not break CI pipelines is to enable the fix flag.
```

**Comment by owenwaller** [comment]
As can been seen in my comment in issue #2671, which has been closed as its a duplicate of this issue which I hadn't seen, this is still very much a live issue.

https://github.com/golangci/golangci-lint/issues/2671#issuecomment-2310658099

From reading the comments above it looks there is no solution at present.

Is this still the case?

Owen
